ui/helpers/taskcluster.js

```diff
@@ -1,48 +1,104 @@
-import { OIDCCredentialAgent, Queue } from 'taskcluster-client-web';
+import { Queue } from 'taskcluster-client-web';
+import debounce from 'lodash/debounce';
+import moment from 'moment';
 
-import { loginRootUrl, getUserSessionUrl } from './url';
+import { clientId, redirectURI } from '../taskcluster-auth-callback/constants';
+
+import { loginRootUrl, createQueryParams } from './url';
+
+export const tcCredentialsMessage =
+  'Need to retrieve or renew Taskcluster credentials before action can be performed.';
 
 const taskcluster = (() => {
-  let credentialAgent = null;
+  let _rootUrl = null;
+
+  // from the MDN crypto.getRandomValues doc
+  const secureRandom = () =>
+    window.crypto.getRandomValues(new Uint32Array(1))[0] / 4294967295;
+
+  const generateNonce = () => {
+    let value = '';
+    const characters =
+      'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+
+    for (let i = 0; i <= 20; i++) {
+      value += characters.charAt(
+        Math.floor(secureRandom() * characters.length),
+      );
+    }
+
+    localStorage.setItem('requestState', value);
+    return value;
+  };
 
-  // Create an OIDC credential agent if it doesn't exist.
-  const tcAgent = () => {
-    if (credentialAgent) {
-      return credentialAgent;
+  const getAuthCode = debounce(
+    () => {
+      const nonce = generateNonce();
+      localStorage.setItem('requestState', nonce);
+      localStorage.setItem('tcRootUrl', _rootUrl);
```

Comment:
Would storing the rootUrl in an environment variable make more sense? Curious as to why we are using localStorage.

Reply:
It's for use in the `TaskclusterCallback` component since that's needed to fetch the token and credentials. This view - `taskcluster-auth`- has its own route independent of Treeherder and Perfherder, so it doesn't have access to a joint top level component where a variable could be stashed regardless of where the user is in the app when credentials are retrieved. 

I can add a comment to explain why that's being stored here.